### PR TITLE
Adds deterministic seeding for EndModel, submodules; adds hyperparam search over submodules

### PR DESCRIPTION
**Pull request by ajratner**
Major changes:
- Adds tests to ensure that given a random seed, `EndModel` is deterministic, including with submodules, e.g. `LSTMModule` as input layer
- Adds support to `Tuner` classes to search over the submodule hyperparameters
- Sets seed in `LSTMModule` random embeddings initialization

Minor things to check:
- Gets rid of unnamed pass-through hyperparameters in `LSTMModule` (to `nn.LSTM`); this was prompted by making things work in the `Tuner`, but seems much cleaner? Can always just add more explicitly named pass-through kwargs...
- Gets rid of cycling through random seeds in `Tuner` class during model search. The `Tuner` already has a global seed, and the models have seeds, so this just seems unnecessary, unless I am missing something?  It seems natural to expect that if two configs in the search space have the same hyperparams, the models should output the same results.

Other todos:
- [x] Add option to use different `validation_metric` in `Tuner`
- [x] Fix handling of `model.eval` and `model.train`; default to the former, except for in `self._train_model`
- [x] Make sure correct validation metric passed through to dev checkpointer in `Tuner`
- [x] Stop the `Tuner` from double-printing scores when dev checkpointer on

### CODE REVIEW

**Comment by ajratner** [comment]
@jason-fries LMK if this fixes things for you re: Issue #100 !

**Comment by jason-fries** [comment]
Some quick issues (all on CPU for now):
- `validation_metric` doesn't look like it's being respected. I'm using F1, which is printed during each epoch update, but definitely not being printed out in the "best model" summary at the end of each loop. Validation scores also still change.
```
[E:9]	Train Loss: 0.033	Dev score: 0.522
Restoring best model from iteration 5 with score 0.551
Finished Training
Accuracy: 0.809
        y=1    y=2   
 l=1    36     51    
 l=2    26     290   
Accuracy: 0.824
        y=1    y=2   
 l=1    38     47    
 l=2    24     294   
```

- validation scoring during param search should disable gradient history (can't see if this is being done or not now), e.g., something like:
```
end_model.eval()
with torch.set_grad_enabled(False):
    y_pred, y_true, y_proba = end_model._get_predictions(test, return_probs=True)
```
This gives deterministic scoring on CPU for me.

**Comment by ajratner** [comment]
Almost ready for review!

**Comment by ajratner** [comment]
@jason-fries @jdunnmon @bhancock8 ready for review :)!

File: metal/tuners/tuner.py [hunk 1]
```diff
@@ -98,6 +108,23 @@ def _test_model_config(
...
+        # Initialize modules if provided
+        for module_name, module_class in self.module_classes.items():
+
+            # Also integrate generated config into module kwargs so that module
+            # hyperparameters can be searched over as well
+            module_kwargs[module_name] = recursive_merge_dicts(
+                module_kwargs[module_name], config, misses="insert"
+            )
+
+            # Initialize module
```
[the remark this answers is not in the document]
Comment (ajratner) [reply metal/tuners/tuner.py]: No it is initializing :)

File: metal/tuners/tuner.py [hunk 2]
```diff
@@ -14,7 +14,10 @@ class ModelTuner(object):
     """A tuner for models
 
     Args:
-        model: (nn.Module) The model class to train (uninitiated)
+        model_class: (nn.Module class) The model class to train (uninitiated)
+        module_classes: (dict) An optional dictionary of module classes
+        (uninitiated), with keys corresponding to their kwargs in model class,
+        e.g. "input_module" in EndModel.
```
[the remark this answers is not in the document]
Comment (ajratner) [reply metal/tuners/tuner.py]: Edited

**Comment by ajratner** [comment]
@jason-fries @bhancock8 Going to merge in (blocking some expts) but lmk if any comments later on!

**Comment by bhancock8** [comment]
Yes, sorry on the delay in reviewing! I'm staying in a 300-year old riyad in Morroco with the wifi to match. :/ I'll review once I'm back and make any comments.
